ImGuiFileDialog.cpp

```diff
@@ -75,300 +103,330 @@ ImGuiFileDialog::~ImGuiFileDialog()
+    int             n                   = 0;
+
+    m_FileList.clear();
+
+    if (0u == m_CurrentPath_Decomposition.size())
+    {
+        SetCurrentDir(vPath);
+    }
+
+    if (0u != m_CurrentPath_Decomposition.size())
+    {
+        n = scandir(vPath.c_str(), &files, NULL, alphaSort);
+        if (n > 0)
+        {
+            for (i = 0; i < n; i++)
+            {
+                struct dirent *ent = files[i];
+
+                FileInfoStruct infos;
+
+                infos.fileName = ent->d_name;
+                if (("." != infos.fileName) && (".." != infos.fileName))
+                {
+                    switch (ent->d_type)
+                    {
+                        case DT_REG: infos.type = 'f'; break;
+                        case DT_DIR: infos.type = 'd'; break;
+                        case DT_LNK: infos.type = 'l'; break;
+                    }
+
+                    if (infos.type == 'f')
+                    {
+                        size_t lpt = infos.fileName.find_last_of(".");
+                        if (lpt != std::string::npos)
+                        {
+                            infos.ext = infos.fileName.substr(lpt);
+                        }
+                    }
+
+                    m_FileList.push_back(infos);
+                }
+            }
+
+            for (i = 0; i < n; i++)
+            {
+                free(files[i]);
+            }
+            free(files);
+        }
+
+//       std::sort(m_FileList.begin(), m_FileList.end(), stringComparator);
+    }
 }
 
 void ImGuiFileDialog::SetCurrentDir(std::string vPath)
 {
-	DIR *dir = opendir(vPath.c_str());
-	if (dir == 0)
-	{
-		vPath = ".";
-		dir = opendir(vPath.c_str());
-	}
-	if (dir != 0)
-	{
-		std::wstring ws(dir->wdirp->patt);
-		m_CurrentPath = std::string(ws.begin(), ws.end());
-		ReplaceString(m_CurrentPath, "\\*", "");
-		closedir(dir);
-		m_CurrentPath_Decomposition = splitStringVector(m_CurrentPath, '\\');
-	}
+    DIR  *dir                 = opendir(vPath.c_str());
+    char  real_path[PATH_MAX];
+
+    if (NULL == dir)
+    {
+        vPath = ".";
+        dir = opendir(vPath.c_str());
+    }
+
+    if (NULL != dir)
+    {
+        if (NULL != realpath(vPath.c_str(), real_path))
+        {
+            m_CurrentPath = real_path;
+            m_CurrentPath_Decomposition = splitStringVector(m_CurrentPath, PATH_SEP);
+        }
+
+        closedir(dir);
+    }
 }
 
 void ImGuiFileDialog::ComposeNewPath(std::vector<std::string>::iterator vIter)
 {
-	m_CurrentPath = "";
-	while (vIter != m_CurrentPath_Decomposition.begin())
-	{
-		if (m_CurrentPath.size() > 0)
-			m_CurrentPath = *vIter + "\\" + m_CurrentPath;
-		else
-			m_CurrentPath = *vIter;
-		vIter--;
-	}
-
-	if (m_CurrentPath.size() > 0)
-		m_CurrentPath = *vIter + "\\" + m_CurrentPath;
-	else
-		m_CurrentPath = *vIter + "\\";
+    m_CurrentPath = "";
+
+    while (true)
+    {
+        if (!m_CurrentPath.empty())
+        {
+            if (*vIter == s_fs_root)
```

Comment:
It took me a few tries to get this code right. I kept getting multiple / at the start of the path rather than just one.